lib/Transforms/Instrumentation/ShadowMemSeaDsa.cc

```diff
@@ -877,6 +911,227 @@ void ShadowDsaImpl::updateReadMod(Function &F, NodeSet &readSet,
+}
+
+CallInst &ShadowDsaImpl::getParentDef(CallInst &memOp) {
+  assert(memOp.getMetadata(m_metadataTag));
+
+  if (isMemInit(memOp))
+    return memOp;
+
+  auto *fn = memOp.getCalledFunction();
+  assert(fn);
+  assert(fn == m_memLoadFn || fn == m_memStoreFn || fn == m_memTrsfrLoadFn);
+
+  assert(memOp.getNumOperands() >= 1);
+  Value *defArg = memOp.getOperand(1);
+  assert(defArg);
+  assert(isa<CallInst>(defArg));
+  auto *def = cast<CallInst>(defArg);
+  auto *defFn = def->getCalledFunction();
+
+  assert(defFn);
+  assert(defFn == m_memStoreFn || defFn == m_argModFn || isMemInit(*def));
+
+  return *def;
+}
+
+const ShadowDsaImpl::MaybeAllocSites &
+ShadowDsaImpl::getAllAllocSites(Value &ptr, AllocSitesCache &cache) {
+  // Performs a simple walk over use-def chains until it finds all allocation
+  // sites or an instruction it cannot look thru (e.g., a load).
+  assert(ptr.getType()->isPointerTy());
+
+  auto *strippedInit = ptr.stripPointerCastsAndBarriers();
+  {
+    auto it = cache.find(strippedInit);
+    if (it != cache.end())
+      return it->second;
+  }
+
+  SmallDenseSet<Value *, 8> allocSites;
+  DenseSet<Value *> visited;
+  SmallVector<Value *, 8> worklist = {strippedInit};
+
+  while (!worklist.empty()) {
+    Value *current = worklist.pop_back_val();
+    assert(current);
+    Value *stripped = current->stripPointerCastsAndBarriers();
+    if (visited.count(stripped) > 0)
+      continue;
+
+    visited.insert(stripped);
+    LOG("shadow_optimizer", llvm::errs() << "Visiting: " << *stripped << "\n");
+
+    // In our memory model, all function calls returning pointers are considered
+    // allocating.
+    if (isa<AllocaInst>(stripped) || isa<Argument>(stripped) ||
+        isa<GlobalValue>(stripped) || isa<CallInst>(stripped)) {
+      allocSites.insert(stripped);
+      continue;
+    }
+
+    if (auto *gep = dyn_cast<GetElementPtrInst>(stripped)) {
+      worklist.push_back(gep->getPointerOperand());
+      continue;
+    }
+
+    if (auto *phi = dyn_cast<PHINode>(stripped)) {
+      for (Value *v : llvm::reverse(phi->incoming_values()))
+        worklist.push_back(v);
+
+      continue;
+    }
+
+    if (auto *gamma = dyn_cast<SelectInst>(stripped)) {
+      worklist.push_back(gamma->getFalseValue());
+      worklist.push_back(gamma->getTrueValue());
+      continue;
+    }
+
+    LOG("shadow_optimizer",
+        llvm::errs() << "Cannot retrieve all alloc sites for " << ptr
+                     << "\n\tbecause of the instruction: " << *stripped
+                     << "\n");
+    return (cache[strippedInit] = llvm::None);
+  }
+
+  return (cache[strippedInit] = allocSites);
+}
+
+bool ShadowDsaImpl::mayClobber(CallInst &memDef, CallInst &memUse, AllocSitesCache &cache) {
+  LOG("shadow_optimizer", llvm::errs() << "\n~~~~\nmayClobber(" << memDef
+                                       << ", " << memUse << ")?\n");
+
+  // Right now memUse must be a load instruction.
```

Comment:
what is the meaning of this comment?

How are uses by functions is handled?

Reply:
uses by functions are not handled right now -- only the simplest case with loads is supported